File: frontend/src/features/reports/lib/permissions.ts
function has(permissions: string[] | undefined, codename: string): boolean {
  if (!permissions) return false
  return permissions.includes('*') || permissions.includes(codename)
}

export function canViewSalesReport(permissions: string[] | undefined): boolean {
  return has(permissions, 'sales.view_sale')
}

export function canViewInstallmentReport(permissions: string[] | undefined): boolean {
  return has(permissions, 'installments.view_installment')
}

export function canViewLeadReport(permissions: string[] | undefined): boolean {
  return has(permissions, 'crm.view_lead')
}

export function canViewLandInventoryReport(permissions: string[] | undefined): boolean {
  return has(permissions, 'plots.view_plot')
}

export function canViewCashFlowReport(permissions: string[] | undefined): boolean {
  return has(permissions, 'finance.view_journalline')
}

File: backend/apps/reports/permissions.py
from rest_framework.permissions import BasePermission

from apps.finance.permissions import HasFinanceReportAccess

__all__ = [
    'HasSalesReportAccess', 'HasInstallmentReportAccess', 'HasLeadReportAccess',
    'HasLandInventoryReportAccess', 'HasFinanceReportAccess',
]


class _HasReportAccess(BasePermission):
    """Reports have no model of their own — gate on the view permission of
    the underlying data they aggregate, same permission a user would already
    need to browse that data directly."""

    required_perm = None

    def has_permission(self, request, view):
        return bool(
            request.user and request.user.is_authenticated and request.user.has_perm(self.required_perm),
        )


class HasSalesReportAccess(_HasReportAccess):
    required_perm = 'sales.view_sale'


class HasInstallmentReportAccess(_HasReportAccess):
    required_perm = 'installments.view_installment'


class HasLeadReportAccess(_HasReportAccess):
    required_perm = 'crm.view_lead'


class HasLandInventoryReportAccess(_HasReportAccess):
    required_perm = 'plots.view_plot'
